internal/stream/stream_format.go

```diff
@@ -94,6 +107,33 @@ func (sf *streamFormat) writeUnitInner(s *Stream, medi *description.Media, u uni
 
 	atomic.AddUint64(s.bytesReceived, size)
 
+	if sf.gopCache && medi.Type == description.MediaTypeVideo {
+		if isKeyFrame(u) {
+			if s.CachedUnits == nil {
+				// Empty the cache and enable caching
+				s.CachedUnits = []unit.Unit{}
+			} else {
+				// Keep the last packets that were used to generate the key frame. This is to send a full key frame in the RTSP stream.
+				lastPTS := s.CachedUnits[len(s.CachedUnits)-1].GetPTS()
+				i := len(s.CachedUnits)
+				for ; i > 0; i-- {
+					if s.CachedUnits[i-1].GetPTS() != lastPTS {
+						break
+					}
+				}
+				s.CachedUnits = s.CachedUnits[i:]
+			}
+
+		}
+		if s.CachedUnits != nil {
+			s.CachedUnits = append(s.CachedUnits, u)
+		}
+		l := len(s.CachedUnits)
+		if l > 512 {
+			s.CachedUnits = s.CachedUnits[l-512:]
```

Comment:
So, we want to keep the cache to a max size of 512 and we truncate the beginning if it grows.
If we only care about 512 items, ideally we shouldn't allocate memory multiple times. I believe the `append` call on line 129 can make allocations when the underlying array needs to grow. 

This seems like a good use case for a ring buffer. How about this one?
https://pkg.go.dev/container/ring

Reply:
`append` will reallocate multiple arrays when it grows but if you know the size, when you allocate it, you can do `make([]s.CachedUnits, 0, 512)` and the when you want to reset it: `s.CachedUnits = s.CachedUnits[:0]`. Then calling append will use the existing underlying array.

Also I had no idea there was a ring buffer in the standard lib!

Reply:
I'll keep this in mind!

Reply:
I changed it to `s.CachedUnits = make([]unit.Unit, 0, MaxCachedGOPSize)`. I will try the ring later, and monitor the memory impact.